lib/sql.ts

```diff
@@ -0,0 +1,147 @@
+/**
+ * Async worker thread wrapper around SQLite, written to improve concurrent performance.
+ * @author mia-pi-git
+ */
+import {Worker, parentPort, isMainThread, workerData} from 'worker_threads';
+import * as Sqlite from 'better-sqlite3';
+
+interface SQLOptions {
+	file: string;
+	/** file to import database functions from - this should be relative to this filename. */
+	extension?: string;
+}
+
+type InputType = unknown[] | Record<string, unknown>;
+
+export type Message = {
+	type: 'prepare', data: string,
+} | {
+	type: 'all', data: InputType, num: number,
+} | {
+	type: "exec", data: string,
+} | {
+	type: 'get', data: InputType, num: number,
+} | {
+	type: 'run', data: InputType, num: number,
+};
+
+if (!isMainThread) {
+	let statementNum = 0;
+	const statements: Map<number, Sqlite.Statement> = new Map();
+	const {file, extension} = workerData;
+	const database = new Sqlite(file);
+	if (extension) {
+		// eslint-disable-next-line @typescript-eslint/no-var-requires
+		const functions: {[k: string]: (...params: any) => any} = require(`../${extension}`);
+		for (const key in functions) {
+			database.function(key, functions[key]);
+		}
+	}
+	database.pragma(`foreign_keys=on`);
+	parentPort!.on('message', (message: Message) => {
+		let statement;
+		let results;
+		switch (message.type) {
+		case 'prepare': {
+			const {data} = message;
+			const nextNum = statementNum++;
+			try {
+				statements.set(nextNum, database.prepare(data));
+			} catch (e) {
+				return console.log(e);
+			}
+			return parentPort!.postMessage(nextNum);
+		}
+		case 'all': {
+			const {num, data} = message;
+			statement = statements.get(num);
+			results = statement?.all(data);
+		}
+			break;
+		case 'get': {
+			const {num, data} = message;
+			statement = statements.get(num);
+			results = statement?.get(data);
+		}
+			break;
+		case 'run': {
+			const {num, data} = message;
+			statement = statements.get(num);
+			results = statement?.run(data);
+		}
+			break;
+		case 'exec': {
+			const {data} = message;
+			database.exec(data);
+		}
+			break;
+		}
+		parentPort!.postMessage(results);
+	});
+}
+
+export class SQLDatabase extends Worker {
```

Comment:
We've had a lot of trouble extending builtins. I think wrapping the builtin makes much more sense.